hibernate-core/src/main/java/org/hibernate/engine/internal/Cascade.java

```diff
@@ -102,25 +100,46 @@ public static void cascade(
 					final Object child;
 					if ( isUninitializedProperty  ) {
 						// parent is a bytecode enhanced entity.
-						// cascading to an uninitialized, lazy value.
+						// Cascade to an uninitialized, lazy value only if
+						// parent is managed in the PersistenceContext.
+						// If parent is a detached entity being merged,
+						// then parent will not be in the PersistencContext
+						// (so lazy attributes must not be initialized).
+						if ( eventSource.getPersistenceContext().getEntry( parent ) == null ) {
+							// parent was not in the PersistenceContext
+							continue;
+						}
 						if ( types[ i ].isCollectionType() ) {
-							// The collection does not need to be loaded from the DB.
-							// CollectionType#resolve will return an uninitialized PersistentCollection.
-							// The action will initialize the collection later, if necessary.
-							child = types[ i ].resolve( LazyPropertyInitializer.UNFETCHED_PROPERTY, eventSource, parent );
-							// TODO: it would be nice to be able to set the attribute in parent using
-							// persister.setPropertyValue( parent, i, child ).
-							// Unfortunately, that would cause the uninitialized collection to be
-							// loaded from the DB.
+							// CollectionType#getCollection gets the PersistentCollection
+							// that corresponds to the uninitialized collection from the
+							// PersistenceContext. If not present, an uninitialized
+							// PersistentCollection will be added to the PersistenceContext.
+							// The action may initialize it later, if necessary.
+							// This needs to be done even when action.performOnLazyProperty() returns false.
+							final CollectionType collectionType = (CollectionType) types[i];
+							child = collectionType.getCollection(
+									collectionType.getKeyOfOwner( parent, eventSource ),
+									eventSource,
+									parent,
+									null
```

Comment:
Calling `CollectionType#getCollection` means we can remove the call to `Type#resolve` with `LazyPropertyInitializer.UNFETCHED_PROPERTY` as an argument.